<a href="https://colab.research.google.com/github/thfruchart/1nsi-2020/blob/master/Chap10/Fractions_binaires.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Fraction binaire,  dans différentes bases

Une **fraction binaire** est le quotient d'un nombre entier par une puissance de 2. 

Dans la suite on notera une fraction binaire sous la forme : $f =  \frac{n}{2^k}$ avec n et k entiers naturels

## Exemples en base 10

In [1]:
f1 = 123 / 2**5
print(f1)

3.84375


In [2]:
f2 =2021/2**10
print(f2)

1.9736328125


## Ecriture de f1 en base 2 (binaire)

Pour un nombre écrit en binaire, diviser par $2^n$ revient à "décaler la virgule de $n$ places vers la gauche. 

In [3]:
bin(123)

'0b1111011'

$f1 =  \frac{123}{2^5}$  peut donc s'obtenir en décalant la virgule de 5 places vers la gauche, dans le nombre $123 = (1111011)_2$

On obtient :

$f1 = \frac{123}{2^5} = (3.84375)_{10} = (11,11011)_2$

#### Vérification :

Chaque chiffre binaire représente une puissance de 2
* pour la partie entière, l'exposant est positif ou nul
* pour la partie fractionnaire, l'exposant est négatif

|poids|$2^1$|$2^0$||$2^{-1}$|$2^{-2}$|$2^{-3}$|$2^{-4}$|$2^{-5}$|
|:--|:--:|:--:||:--:|:--:|:--:|||
|chiffres binaires|1|1|,|1|1|0|1|1|

In [4]:
2**1 + 2**0 + 2**-1 + 2**-2 +2**-4 + 2**-5

3.84375

## Ecriture de f1 en base 16 (hexadécimal)

On peut convertir un nombre écrit en binaire, vers la base 16 en groupant les chiffre par **bloc de quatre** à partir de la virgule : 

$f1 =  (11,11011)_2 =  (0011,1101 \ 1000)_2$

Chaque bloc de quatre chiffre binaire correspond à un chiffre hexadécimal : 

$f1 =   (0011,1101 \ 1000)_2 = ( ... ... ...)_{16}$

#### Vérification

Le type `float` possède une méthode `fromhex()` qui prend en argument un chaîne de caractère décrivant un flottant en hexadécimal. 

In [6]:
float.fromhex('0x3.D8')

3.84375

## Activité

In [7]:
f2 =2021/2**10
print(f2)

1.9736328125


En suivant l'exemple donné pour f1 : 
1. écrire la fraction binaire f2 en base 2
2. écrire la fraction binaire f2 en base 16
3.  puis vérifier avec le type flottant de Python

In [8]:
bin(2021)

'0b11111100101'

1. l'écriture binaire de f2 s'obtient en décalant la virgule de 10 places (division par $2^{10}$)

donc f2 =`1,1111100101`

2. l'écriture binaire de f2 en hexadécimal : on groupe les chiffres par quartet
donc 

f2 =`0001,1111 1001 0100`

f2 =`1, F 9 4` en hexadécimal

In [9]:
#3. 
float.fromhex('0x1.f94')

1.9736328125

In [10]:
f2

1.9736328125

# Nombre décimal et fraction binaire

On peut convertir facilement le nombre f1 en binaire, d'après son écriture décimale : `3.84375`
* la partie entière possède une écriture binaire "ordinaire" : $(3)_{10} = (11)_2$
* on trouve les chiffres de la partie fractionnaire `0.84375` avec la méthode des multiplications par 2 de la succession des parties fractionnaires :  

In [12]:
0.84375 * 2

1.6875

In [13]:
0.6875*2

1.375

In [14]:
0.375*2

0.75

In [15]:
0.75*2

1.5

In [16]:
0.5*2

1.0

In [17]:
0*2

0

**1**.6875

**1**.375

**0**.75

**1**.5

**1**.0

**0**.0

donne pour `0.84375`  l'écriture en binaire :  **`0.110110`**

## Activité

Avec la même méthode, donner une écriture binaire des nombres décimaux : 

$a = 0,55078125$ : on trouve `0.10001101`

$b = 0,2$ : **problème** on trouve b environ égal à `0.0011 0011 0011 ... `

$c = \frac{1}{3}$ : **problème** on trouve c environ égal à  `0.010101010101...`


Calcul de a

In [18]:
0.55078125*2

1.1015625

In [19]:
0.1015625*2

0.203125

In [21]:
0.203125*2

0.40625

In [22]:
0.40625*2

0.8125

In [23]:
0.8125*2

1.625

In [24]:
0.625*2

1.25

In [25]:
0.25*2

0.5

In [26]:
0.5*2

1.0